docs/developers/testing.md

```diff
@@ -0,0 +1,314 @@
+
+The Python `unittest`-based test suite known as the "core" tests is a required verification step in order to release the iRODS server. Verifying the iRODS server before release involves running the core tests in the following configurations for all supported operating systems, and database types and versions:
+
+ - catalog service provider
+ - topology on catalog service provider (with and without SSL enabled)
+ - topology on catalog service consumer (with and without SSL enabled)
+
+The list of tests can be found here: [https://github.com/irods/irods/blob/main/scripts/core_tests_list.json](https://github.com/irods/irods/blob/main/scripts/core_tests_list.json)
+
+Additionally, there is a set of tests having to do specifically with federated zones which must be passing as well.
+
+### How to run the tests
+
+The [iRODS Testing Environment](#irods-testing-environment) is the primary tool for verifying the iRODS server software. If you do not want to use the iRODS Testing Environment, you must install the iRODS server packages on some host and run the following command **as the iRODS service account user**:
+```
+python /var/lib/irods/scripts/run_tests.py
+```
+This serially runs the entire test suite on the local host. The `--run_specific_tests` option enables the user to specify a dot-delimited test name to run. The test file and test class specified must exist in `/var/lib/irods/scripts/irods/test` for this to work.
+
+### How to add tests
+
+The Python tests are maintained in the main iRODS repository. The test files are located in the `scripts/irods/test` directory. The Python test suite is based on the `unittest` test framework. To learn more about this, see the documentation: [https://docs.python.org/3/library/unittest.html#module-unittest](https://docs.python.org/3/library/unittest.html#module-unittest)
+
+Remember: Every test file, test class, and test case must have a name which begins with the string `test_`. This is required by Python's `unittest` test discovery mechanism.
+
+#### Adding a test
+
+Adding a new test file, test class, or test case is a matter of discretion on the part of the developer. Here are some hints for helping to make the decision:
+
+ 1. If you are able to locate an existing test class which is related to the feature or issue being tested and provides all the setup required by your use case, simply add a new method to the existing test class.
+ 2. If you are able to locate an existing file which seems related to the feature or issue being tested but your use case requires a specific kind of setup, add a new test class to the existing test file.
+ 3. If you are unable to find an existing file which seems related to the feature or issue being tested, or you wish to keep your tests separated, add a new test file with its own test classes.
+ 4. If you are able to locate an existing test case which is precisely related to the feature or issue being tested, modify the existing test case to cover the new functionality.
```

Comment:
feels like these four are not in either an ascending or descending order of specificity...  please revisit/consider.

Reply:
I guess the ascending order of specificity would go like this:

a. Adding a new test file
b. Adding a new test class to an existing test file
c. Adding a new test case to an existing test class
d. Modifying an existing test case

So the mapping right now seems to be...
 - 1 -> c
 - 2 -> b
 - 3 -> a
 - 4 -> d

And we want it to be...
 - 1 -> a
 - 2 -> b
 - 3 -> c
 - 4 -> d

So the overall change would be flipping items 1 and 3. Does that seem right?

Reply:
yes, i think i agree, flip 1 and 3..

and then also possibly, flip the entire listing so it's specific (smallest edit) to general (biggest edit)?...
d, c, b, a

Reply:
I prefer option 4 being where it is so that it's not as prominent. I almost contemplated not mentioning it at all, but it felt incomplete without it. I can reverse the order if desired, though.

Reply:
there is a similar listing below - please eyeball that one for ordering too.

Reply:
Similar edits made to list down there. Apparently there were two '1's as well. Oops.

Reply:
ha, missed that.